desmume/src/saves.cpp

```diff
@@ -744,25 +748,37 @@ void loadstate_slot(int num)
 			static const char PSS = '/';
 			#endif
 			mkdir(dirname.c_str(),0777);
-			static unsigned seed;
-			for(;;)
+			
+			int cur_index = -1; // setup index, -1 in case it is first instance
+			if (max_index == -1) { // If not set
+				max_index = 200;
+			}
+			
+			std::string index_fname = dirname + PSS + "backup.index";
+			FILE* index_file = fopen(index_fname.c_str(), "r+"); // Read/update but don't create
+			if (index_file) {
+				fscanf(index_file, "%d", &cur_index);
+				rewind(index_file); // prepare to overwrite
+			}
+			else
 			{
-				std::string fname = dirname + PSS;
-				char mini[100];
-				sprintf(mini,"%u",seed);
-				fname += mini + (std::string)".dst";
-				FILE* f = fopen(fname.c_str(),"rb");
-				if(f)
-				{
-					seed = rand()*16000+rand();
-					fclose(f);
-					continue;
-				}
-				seed++;
-				savestate_save(fname.c_str());
-				printf("Creating backup of current state prior to loadstate as path: %s\n",fname.c_str());
-				break;
+				index_file = fopen(index_fname.c_str(), "w"); // Create if doesn't exist
```

Comment:
this may return 0 too for whatever reason, leading to a crash later on

Reply:
Good point.
I added check to make sure that index was opened, and if not skip backup part.
I'm assuming it is still preferable to proceed with loading state even if backup can not be created.